kernel/arch/dreamcast/hardware/cdrom.c

```diff
@@ -409,12 +860,54 @@ void cdrom_init(void) {
     /* Reset system functions */
     syscall_gdrom_reset();
     syscall_gdrom_init();
+
+    unlock_dma_memory();
     mutex_unlock(&_g1_ata_mutex);
 
+    /* Hook all the DMA related events. */
+    old_dma_irq = asic_evt_set_handler(ASIC_EVT_GD_DMA, g1_dma_irq_hnd, NULL);
+    asic_evt_set_handler(ASIC_EVT_GD_DMA_OVERRUN, g1_dma_irq_hnd, NULL);
+    asic_evt_set_handler(ASIC_EVT_GD_DMA_ILLADDR, g1_dma_irq_hnd, NULL);
+
+    if(old_dma_irq.hdl == NULL) {
+        asic_evt_enable(ASIC_EVT_GD_DMA, ASIC_IRQB);
+        asic_evt_enable(ASIC_EVT_GD_DMA_OVERRUN, ASIC_IRQB);
+        asic_evt_enable(ASIC_EVT_GD_DMA_ILLADDR, ASIC_IRQB);
+    }
+
+    vblank_hnd = vblank_handler_add(cdrom_vblank, NULL);
+    inited = true;
+
     cdrom_reinit();
 }
 
 void cdrom_shutdown(void) {
 
-    /* What would you want done here? */
+    if(!inited) {
+        return;
+    }
+
+    vblank_handler_remove(vblank_hnd);
+
+    /* Unhook the events and disable the IRQs. */
+    if(old_dma_irq.hdl) {
+        /* G1-ATA driver uses the same handler for 3 events. */
+        asic_evt_set_handler(ASIC_EVT_GD_DMA,
+            old_dma_irq.hdl, old_dma_irq.data);
+        asic_evt_set_handler(ASIC_EVT_GD_DMA_OVERRUN,
+            old_dma_irq.hdl, old_dma_irq.data);
+        asic_evt_set_handler(ASIC_EVT_GD_DMA_ILLADDR,
+            old_dma_irq.hdl, old_dma_irq.data);
+
+        old_dma_irq.hdl = NULL;
```

Comment:
Why do you need to restore the handlers btw? This is `cdrom_shutdown()`, nothing is really supposed to use the GD IRQs after that point (until cdrom_init() is called again).

Reply:
I'm restoring them for G1 ATA, not for GD-drive.
It's just that the API for ASIC IRQ handlers didn't assume the use of the same interrupts in different places (only vblank). I've modified it a bit to make it possible, but it looks like this, yes.